snuba/web/split.py

```diff
@@ -128,61 +174,100 @@ def time_split(dataset, request: Request, *args, **kwargs):
+    build the full result set.
+    """
+
+    def __init__(
+        self, id_column: str, project_column: str, timestamp_column: str
+    ) -> None:
+        self.__id_column = id_column
+        self.__project_column = project_column
+        self.__timestamp_column = timestamp_column
+
+    def __get_min_columns(self) -> Sequence[str]:
+        return [self.__id_column, self.__project_column, self.__timestamp_column]
+
+    def __can_execute(self, request: Request) -> bool:
+        (use_split,) = state.get_configs([("use_split", 0)])
+        if not use_split:
+            return False
+
+        if not _is_query_splittable(request.query):
+            return False
+
+        if not request.query.get_selected_columns():
+            return False
+
+        if request.query.get_aggregations():
+            return False
+
+        total_col_count = len(request.query.get_all_referenced_columns())
+        copied_query = copy.deepcopy(request.query)
+        copied_query.set_selected_columns(self.__get_min_columns())
+        min_col_count = len(copied_query.get_all_referenced_columns())
+
+        return total_col_count > min_col_count
+
+    def execute(
+        self, request: Request, runner: QueryRunner
+    ) -> Optional[RawQueryResult]:
         """
         Split query in 2 steps if a large number of columns is being selected.
             - First query only selects event_id and project_id.
             - Second query selects all fields for only those events.
             - Shrink the date range.
         """
+        if not self.__can_execute(request):
+            return None
+
         # The query function may mutate the request body during query
         # evaluation, so we need to copy the body to ensure that the query has
         # not been modified by the time we're ready to run the full query.
         minimal_request = copy.deepcopy(request)
-        minimal_request.query.set_selected_columns(column_split_spec.get_min_columns())
-        result = query_func(dataset, minimal_request, *args, **kwargs)
+        minimal_request.query.set_selected_columns(self.__get_min_columns())
+        result = runner(minimal_request)
         del minimal_request
 
         if result.result["data"]:
             request = copy.deepcopy(request)
 
             event_ids = list(
-                set(
-                    [
-                        event[column_split_spec.id_column]
-                        for event in result.result["data"]
-                    ]
-                )
-            )
-            request.query.add_conditions(
-                [(column_split_spec.id_column, "IN", event_ids)]
+                set([event[self.__id_column] for event in result.result["data"]])
             )
+            request.query.add_conditions([(self.__id_column, "IN", event_ids)])
             request.query.set_offset(0)
             request.query.set_limit(len(event_ids))
```

Comment:
Just noticed this assumes the `event_id` is globally unique rather than unique per project, so this logic is incorrect but probably not often enough to be noticable.

Reply:
Added a TODO.

Reply:
Good point. I don't recall why that `set()` was needed.